### PR TITLE
Fix out= variant recompilations for CompositeImplicitAutograd linalg ops

### LINKED ISSUE #187953: torch.compile: out= variants of CompositeImplicitAutograd linalg ops (linalg.norm, cholesky, ...) recompile on every input shape
### Describe the bug

The `out=` variants of several `CompositeImplicitAutograd` linalg ops recompile
under `torch.compile` on every new input shape, while the functional variants
specialize only once. Reported originally as part of #135859 for `linalg.norm`
and `cholesky`; the `max` / `min` / `topk` cases (structured ops) are fixed in
#187952, but the composite ops need a different fix.

Minimal repro:

```python
import torch
torch._logging.set_logs(recompiles=True)

def f(x):
    out = x.new_empty(x.shape[0])
    torch.linalg.norm(x, dim=1, out=out)
    return out

cf = torch.compile(f)
for n in range(4, 10):
    cf(torch.randn(n, 6))   # recompiles on every n
```

`torch.linalg.norm(x, dim=1)` (no `out=`) does not recompile. `torch.cholesky`
/ `torch.linalg.cholesky` with `out=` behave the same way.

### Root cause (observed)

The composite `out=` kernel computes the functional result, then resizes the
user-provided out tensor, e.g. `linalg_norm_out` in
`aten/src/ATen/native/LinearAlgebra.cpp`:

```cpp
auto out = at::linalg_norm(...);              // functional: decomposes symbolically, fine
at::native::resize_output(result, out.sizes());   // specializes
result.copy_(out);
```

Under `FakeTensor`, `resize_output` dispatches `aten.size.default`, whose handler
materializes symbolic sizes to Python ints
(`torch/_subclasses/fake_tensor.py`, `_DISPATCH_META_HANDLERS`). Calling `int()`
on the `SymInt` adds an `Eq(s, 4)` guard and specializes the input dim, so the
frame recompiles whenever the shape changes.

The same `compute functional -> resize_output(result, tmp.sizes()) -> copy_`
pattern is used by many composite linalg `out=` kernels (cholesky, qr, eig, eigh,
svd, lstsq, ...), so they likely all share this behavior.

For structured ops (max/min/topk) the fix is to register a meta for the out
overload so FakeTensor never runs the C++ out= path (#187952). That approach is
not available here: registering a meta for a `CompositeImplicitAutograd` op is
explicitly disallowed (`activate_meta()` raises "we shouldn't register meta
function for it").

Possible directions (for maintainers to weigh):
- make `resize_output` / the `aten.size.default` fake handling symbolic-safe when
  the sizes already match, or
- trace these composite `out=` ops via their functional decomposition plus a
  symbolic out copy (the `out_wrapper` `_maybe_resize_out` path), instead of the
  C++ out= kernel.

### Error logs

```
[__recompiles] Recompiling function f ...
    triggered by the following guard failure(s):
    - 0/0: tensor 'x' size mismatch at index 0. expected 4, actual 5
```

### Versions

PyTorch built from source at `main` (2.14.0a0+git37c67b7), Python 3.12,
CUDA 12.8, Windows 11. Also reproduces on CPU.

---

Filed with the assistance of an AI coding assistant (Claude); reviewed by the author.

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano @chauhang @penguinwu @eellison @aorenste @ezyang @bobrenjc93 @aditvenk @laithsakka @oulgen @jamesjwu @anijain2305 @masnesral @coconutruben @bdhirsh

### PR DESCRIPTION
**Pull request by jonasvq**
## Issue

Fixes #187953

Related: #135859 (parent issue) and #187952, which fixed the same
class of `out=` over-recompilation for `torch.max`/`min`/`topk` and explicitly
deferred the CompositeImplicitAutograd linalg `out=` overloads handled here.

## Summary

Under `torch.compile` with dynamic shapes, the `out=` variants of
CompositeImplicitAutograd linalg ops (`linalg.norm`, `linalg.cond`,
`linalg.pinv`, `cholesky_solve`, `cholesky_inverse`, `geqrf`,
`householder_product`, `ormqr`, `linalg_eig`, `linalg_eigvals`, `lstsq`,
`tensorinv`, `tensorsolve`, ...) recompiled on every new input shape, while the
functional variants specialize once. Past the default `recompile_limit` they
fall back to eager entirely.

These kernels compute a functional result into a temporary and write it into the
user `out` tensor via `resize_output(result, tmp.sizes())` + `copy_`. Two things
specialized the symbolic input dim during FakeTensor tracing:

1. `tmp.sizes()` returns `IntArrayRef`; on a FakeTensor with symbolic sizes this
   dispatches `aten.size.default`, whose meta handler materializes every `SymInt`
   with `int(s)`, emitting an `Eq(s0, N)` guard.
2. `_resize_output_check` compared shapes with `ArrayRef::equals`, i.e.
   `SymInt::operator==` -> `guard_bool`, which is not size-oblivious and also
   specializes.

Fix: the kernels pass `tmp.sym_sizes()` (no materialization) to the existing
`resize_output_symint`, and `_resize_output_check` now compares shapes through a
new size-oblivious helper `_shapes_definitely_equal`. The `IntArrayRef` overload
keeps the exact `a.equals(b)` behavior (eager path byte-identical); the
`SymIntArrayRef` overload uses `TORCH_GUARD_OR_FALSE(a[i].sym_eq(b[i]))`,
returning "resize" (a runtime no-op when the shapes match) when equality cannot
be proven instead of guarding. Only the symbolic/compile path changes and it is
strictly conservative.

Alternatives considered: registering the Python refs as decompositions (only
covers `linalg.norm`, changes the matrix spectral/nuclear graph via `prims.svd`,
and does not cover the factorizations, which have no refs); meta registration
(disallowed for CompositeImplicitAutograd ops); making the `aten.size.default`
fake handler symbolic-safe (unbounded blast radius across every `.sizes()`
caller). The chosen fix is localized to the resize machinery plus per-kernel call
sites and fixes the whole family at once.

### Performance

Eager is unaffected (resize-check helper, RTX 4060 Ti, ms/call, out= vs
functional+copy which isolates the resize machinery): `linalg.norm` -12.0%,
`linalg.cond` -1.6%, `cholesky_inverse` +5.7% -- all small-op jitter around
zero. Compile (sweep shapes 4..9, `dynamic=None`): `linalg.norm` and
`linalg.cond` `out=` recompiles drop from 6 (one per shape, then eager fallback
past `recompile_limit`) to 2, matching the functional variant.

## Checklist

- [x] Passes lint (`lintrunner -a`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [x] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

---

This change was authored with the assistance of an AI coding assistant (Claude).
The author reviewed the change and is responsible for it.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?
